docs/proposals/set-ref.md

```diff
@@ -0,0 +1,144 @@
+
+Main goal of *setReference* effort is to solve a problem of overlapping controllers that fight over some resources (e.g. *ReplicaSets* fighting with *ReplicationControllers* over *Pods*), which cause serious [problems](https://github.com/kubernetes/kubernetes/issues/24433) such as exploding memory of Controller Manager.
+
+We don’t want to have (just) an in-memory solution, as we don’t want a Controller Manager crash to cause massive changes in object ownership in the system. I.e. we need to persist the information about "owning controller".
+
+Secondary goal of this effort is to improve performance of various controllers and schedulers, by removing the need for expensive lookup for all matching "controllers".
+
+# Non goals
+
+Cascading deletion is not a goal of this effort. Cascading deletion will use *ownerReferences*, which is a [separate effort](garbage-collection.md).
+
+There will be some interactions between set and owner references, they will have a similar orphaning/adoption behavior and *setReference* may be treated as one of *ownerReferencess*.
+
+# API and semantic changes
+
+There will be a new API field in the *ObjectMeta* in which we will store a *setReference*:
+
+```
+ObjectMeta {
+    …
+    setReference *SetReference
+    …
+}
+```
+
+```
+type SetReference struct {
+    // Version of the referent in group/version format.
+    APIVersion string
+    // Kind of the referent.
+    Kind string
+    // Name of the referent.
+    Name string
+    // UID of the referent.
+    UID types.UID
+    // should this SetReference be treated as a parent for cascading deletion
+    isParent bool
+}
+```
+
+Most controllers (all that manage collections of things defined by label selector) will have slightly changed semantics: currently controller owns an object if its selector matches object’s labels, but after introduction of *setReference* a controller will own an object iff selector matches labels and *setReference* points to it.
+
+If the owner's selector or owned object's labels change, the owning controller will be responsible for orphaning (clearing *setReference*) objects, after which adoption procedure (setting *setReference*) might occur, if another controller has a selector matching..
+
+User shouldn't set *setReference* by himself/herself, only system components should be responsible for updating this field. In the fullness of time an Admission Control plugin may block all updates to *setReference* signed by something different than Controller Manager.
+
+For debugging purposes we want to add an `adoptionTime` annotation which will keep the time of last ownership transfer.
+
+# Upgrade/downgrade procedure
+
+During system upgrade controllers will adopt objects that match their label selector using ordinary adoption procedure. Adoption, by default, will set *IsParent* to *false,* so the behavior of objects existing before the update won’t change (i.e. deleting owners won't cause cascading deletion - see OwnerReference design).
```

Comment:
I think by default *IsParent* should set *true*, so if the user later want to do cascading deletion, it will be possible.

We can keep the backward compatibility by defaulting the DeleteOptions.OrphanDependents to true.